src/lib/schedule.c

```diff
@@ -196,41 +197,54 @@ static struct task *schedule_edf(void)
 	} else {
 		/* yes, run current task */
 		task->start = current;
+
+		/* init task for running */
+		wait_init(&task->complete);
+		spin_lock_irq(&sch->lock, flags);
 		task->state = TASK_STATE_RUNNING;
+		list_item_del(&task->list);
+		spin_unlock_irq(&sch->lock, flags);
+
+		/* now run task at correct run level */
 		arch_run_task(task);
 	}
 
 	/* tell caller about future task */
 	return future_task;
 }
 
-#if 0 /* FIXME: is this needed ? */
-/* delete task from scheduler */
-static int schedule_task_del(struct task *task)
+/* cancel and delete task from scheduler - won't stop it if already running */
+int schedule_task_cancel(struct task *task, int wait)
 {
 	uint32_t flags;
 	int ret = 0;
 
 	tracev_pipe("del");
 
-	/* add task to list */
-	spin_lock_irq(&sch->lock, flags);
-
-	/* is task already running ? */
-	if (task->state == TASK_STATE_RUNNING) {
-		ret = -EAGAIN;
-		goto out;
+	/* check current task state */
+	switch (task->state) {
+	case TASK_STATE_QUEUED:
+		/* delete task */
+		spin_lock_irq(&sch->lock, flags);
+		task->state = TASK_STATE_CANCEL;
+		list_item_del(&task->list);
+		spin_unlock_irq(&sch->lock, flags);
+		break;
+	case TASK_STATE_RUNNING:
+		/* already running, nothing we can do about it atm */
+		if (wait) {
```

Comment:
We need to keep the lock held over this whole function, but we can release it in if (wait) {}

Reply:
Yes, I will update the PR